Add MADX Input Parser

n01r opened the pull request:
This PR adds a first MAD-X (Methodical Accelerator Design) input capability to ImpactX (#104).
We use the [pymadxparser](https://gitlab.psi.ch/adelmann/pymadxparser) that Andreas Adelmann (PSI) generously made available for us. 🎉 

- [x] create free-standing function `madx_to_impactx`
- [x] import MADX parser and translation function correctly with `__init__.py`
- [x] make FODO run with MADX
- [x] make the automated FODO test work (still doesn't find the input file)
- [x] create standards for MADX' available particle species and read charge and mass from these?

### Follow-up features

- parse Fortran number formats (e.g. `40.d-3`)
- implement missing elements
- make inline calculations work
- create a useful intermediate representation that can be exported to other formats
- visualize the lattice representation

n01r committed:
Add MADX Input Parser

Using Andreas Adelmann's pymadxparser

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on examples/CMakeLists.txt:
```diff
@@ -119,6 +119,16 @@ add_impactx_test(FODO.py
 
+# Python MADX: FODO Cell ######################################################
+#
+add_impactx_test(FODO_MADX.py
```
similar to https://github.com/ECP-WarpX/impactx/blob/2d97982240de31a3c27a37f4049dd255ac00ea39/tests/python/CMakeLists.txt#L1-L2
```suggestion
# copy MAD-X lattice file
file(COPY ${ImpactX_SOURCE_DIR}/examples/fodo/input_fodo.madx DESTINATION ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/FODO_MADX.py)

add_impactx_test(FODO_MADX.py
```

We can also write this more generically above into the `add_impactx_test` function, e.g., copying all `*.madx` files found.

n01r committed:
Copy MADX input via CMakeLists.txt

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

n01r committed:
Add beam function to MADX translation

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

n01r commented:
Hey @ax3l, should I make the chicane input look like the one from `input_chicane.in` as well?

Also, the `FODO/chicane_MADX.py.run` and the `FODO/chicane_MADX.py.analysis` are successful but the plot tests are not. Any advice?

ax3l commented:
@n01r as discussed yesterday, can you please adjust the `.madx` file to fit the numerical parameters of the checked in chicane example?

ax3l commented on examples/CMakeLists.txt:
```diff
@@ -119,6 +119,19 @@ add_impactx_test(FODO.py
+# copy MAD-X lattice file
+file(COPY ${ImpactX_SOURCE_DIR}/examples/fodo/input_fodo.madx DESTINATION ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/FODO_MADX.py)
+
+add_impactx_test(FODO_MADX.py
```
```suggestion
add_impactx_test(FODO.MADX.py
```

ax3l commented on examples/CMakeLists.txt:
```diff
@@ -149,6 +162,19 @@ add_impactx_test(chicane.py
+# copy MAD-X lattice file
+file(COPY ${ImpactX_SOURCE_DIR}/examples/chicane/input_chicane.madx DESTINATION ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/chicane_MADX.py)
+
+add_impactx_test(chicane_MADX.py
```
```suggestion
add_impactx_test(chicane.MADX.py
```

ax3l commented on examples/CMakeLists.txt:
```diff
@@ -149,6 +162,19 @@ add_impactx_test(chicane.py
+# Python MADX: Chicane ######################################################
+#
+# copy MAD-X lattice file
+file(COPY ${ImpactX_SOURCE_DIR}/examples/chicane/input_chicane.madx DESTINATION ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/chicane_MADX.py)
```
```suggestion
file(COPY ${ImpactX_SOURCE_DIR}/examples/chicane/chicane.madx DESTINATION ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/chicane.MADX.py)
```

ax3l commented on examples/CMakeLists.txt:
```diff
@@ -119,6 +119,19 @@ add_impactx_test(FODO.py
+# Python MADX: FODO Cell ######################################################
+#
+# copy MAD-X lattice file
+file(COPY ${ImpactX_SOURCE_DIR}/examples/fodo/input_fodo.madx DESTINATION ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/FODO_MADX.py)
```
```suggestion
file(COPY ${ImpactX_SOURCE_DIR}/examples/fodo/fodo.madx DESTINATION ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/FODO.MADX.py)
```

ax3l commented on examples/chicane/input_chicane.madx:
```diff
@@ -0,0 +1,15 @@
+DLEFT:   drift, L=0.7578;
+DMIDDLE: drift, L=0.9575;
+DRIGHT:  drift, L=0.7578;
+! TODO make this work with inline calculations
+! theta=pi/10.;
+! TODO put theta again everywhere after `angle=` with its right sign
```
cc @np-eazy fyi :)

ax3l commented on src/python/impactx/MADXParser.py:
```diff
@@ -0,0 +1,382 @@
+# Author: Matthias Frey, Andreas Adelmann
+# Date: 13. Oct. 2017
+#
+# This file parses MAD-X file into a beamline
+# composed of pyAcceLEGOrator elements.
+#
+# 10.8.2022 Adapted for standalone use
+#
```
Adjust as following, please clarify with Andreas if we can add a simple `BSD-3-Clause` license. Use international date formatting:
```suggestion
#!/usr/bin/env python3
#
# Copyright 2022 ImpactX contributors
# Authors: Matthias Frey, Andreas Adelmann, Marco Garten
# License: BSD-3-Clause
#
# Changelog:
#   Oct 13th, 2017: original version for pyAcceLEGOrator elements
#   Aug 10th, 2022: adapted for standalone use
#
# -*- coding: utf-8 -*-
```

ax3l replied:
Confirmed via email today :)

ax3l commented on src/python/impactx/madx_to_impactx.py:
```diff
@@ -0,0 +1,106 @@
+        _particle = 'generic'
+    else:
+        _particle = particle
+        print('Choosing provided particle species "', particle, '", ignoring potentially provided `charge` and `mass` and setting defaults.')
```
```suggestion
        print(f"Choosing provided particle species "{particle}", ignoring potentially provided `charge` and `mass` and setting defaults.")
```

ax3l committed:
Clean up: Test & Input File Names

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l committed:
MADXParser: Regex Literals

Use proper regex literals.
Avoids:
```
DeprecationWarning: invalid escape sequence \(
```
et al.

ax3l committed:
Lattice: Cleaner User API & Docs

n01r committed:
Fix chicane MADX example and add parser elements

n01r committed:
Add additional parameters to DipEdge

and adjust formatting

n01r committed:
Update examples according to API changes

Caution: Examples still not working again

ax3l committed:
Cleaning

n01r committed:
Rename elements in chicane.madx

to avoid matching of `drift` in `use, sequence` line with parser.

ax3l commented on src/python/impactx/madx_to_impactx.py:
```diff
@@ -0,0 +1,171 @@
+def read_beam(ref: RefPart, madx_file):
+    """
+    Function that reads elements from a MAD-X file into a list of ImpactX.KnownElements
+    :param RefPart ref: ImpactX reference particle (pybind_11 object)
```
```suggestion
    :param RefPart ref: ImpactX reference particle (passed by reference)
```

n01r committed:
Fix: element type instead of name

change MADXParser Quad type to Quadrupole to stay consistent with
MADX input

ax3l committed:
Update Copyright MAD-X Parser Script

Confirmed today via email :)

ax3l reviewed (approved):
:rocket: :sparkles:

ax3l commented on src/python/impactx/MADXParser.py:
```diff
@@ -0,0 +1,470 @@
+                for e in self.__elements:
+                    if elem == e["name"]:
+                        if e["type"] == "drift":
+                            print("Drift L= " + str(e["l"]))
+                            beamline.append(e)
+                        elif e["type"] == "sbend":
+                            print("Sbend L= ", e["l"], " angle = ", e["angle"])
+                            beamline.append(e)
+                        elif e["type"] == "quadrupole":
+                            print("Quadrupole L= ", e["l"], " k1 = ", e["k1"])
+                            beamline.append(e)
+                        elif e["type"] == "dipedge":
+                            print("Dipedge H= ", e["h"], " E1 = ", e["e1"])
```
Leftover debug print statements :)

n01r replied:
Hm, it's there in Andreas' original version, too. Should we completely remove it or make it optional? Otherwise, it will print every beamline element once (however, it's still incomplete, right now).

ax3l replied:
please comment all but the `else:` prints out, this is clearly debug stuff that we don't want to print by default.

ax3l replied:
```suggestion
                            # print("Drift L= " + str(e["l"]))
                            beamline.append(e)
                        elif e["type"] == "sbend":
                            # print("Sbend L= ", e["l"], " angle = ", e["angle"])
                            beamline.append(e)
                        elif e["type"] == "quadrupole":
                            # print("Quadrupole L= ", e["l"], " k1 = ", e["k1"])
                            beamline.append(e)
                        elif e["type"] == "dipedge":
                            # print("Dipedge H= ", e["h"], " E1 = ", e["e1"])
```

ax3l commented on src/python/impactx/madx_to_impactx.py:
```diff
@@ -0,0 +1,172 @@
+                )
+            elif d["type"] == "dipedge":
+                impactx_beamline.append(
+                    elements.DipEdge(
```
CI says:
```
TypeError: __init__(): incompatible constructor arguments. The following argument types are supported:
    1. impactx.impactx_pybind.elements.DipEdge(psi: float, rc: float, g: float, K2: float)

Invoked with: kwargs: psi=0.048345620280243, rc=10.3500000522675, g=0.0, K2=0.0, nslice=25
```

ax3l commented on src/python/impactx/madx_to_impactx.py:
```diff
@@ -0,0 +1,172 @@
+                        # MAD-X is using half the gap height
+                        g=2.0 * d["hgap"],
+                        K2=d["fint"],
+                        nslice=nslice,
```
Thin element
```suggestion
```

ax3l committed:
Fix: DipEdge is a thin element

n01r committed:
Fix Chicane input to make test pass

also fix energy conversion in madx_to_impactx

ax3l committed:
Comment Debug Printing
